feat(reducers): cache reducer for increased performance - @puppybits

puppybits opened the pull request:
This fixes massive performance issues we had using both Firestore & redux-firestore in production. Running a transaction update must round-trip to the server before being reflected in Redux. The mutate + cache reducer allows changes to be reflected in redux synchronously with no additional work. 

There are no breaking changes. But it does adds `id` & `path` properties to all results. Mostly it add the new cache reducer for synchronous optimistic updates. And it adds a new mutate function which the cache reducer uses to reflect changes synchronously. 

- New unit tests for cache reducer & mutate all pass.
- Numerous performance tuning done on the cache reducer for large, complex datasets.
- Handles firestore rejects & offline saving (via onSnapshot.includesMetadataChanges)

### Description

- **Cache Reducer.** UI feels instantaneous to users. 
  - Enables synchronous optimistic updates for new, deleted & updated data. 
  - All data changes synchronously reflected in the cache reducer queries.
  - Once Firestore approves/rejects a change it also removes the optimistic commit.
```js
// selector for cache reducer
(state) => state.firebase.firestore.cache[myStoreAs].docs;

// access a read-only fragment of Firestore based on all active queries
const document = state.firebase.firestore.cache.database[full_path_to_collection][document_id];
```

- **Mutate.** Simplifies saving, batching and transactions. 
  - Easily switch between single saves and atomic transactions. 
  - Simplifies complex saves that have to coordinate through multiple redux actions. 
  - Synchronously updates queries in the cache reducer for the UI to feel immediately responsive to all changes.
```js
// single save
getFirestore().mutate({collection: 'tasks', doc: 'task-1', data: {status: 2}});
// batch
getFirestore().mutate([
  {collection: 'tasks', doc: 'task-1', data: {status: 2}},
  {collection: 'tasks', doc: 'task-2', data: {status: 2}},
]);
// transaction, read results use dependency injection for write functions.
getFirestore().mutate({
  reads: { task1: {collection: 'tasks', doc: 'task-1'} },
  writes: [ ({task1}) => ({collection: 'tasks', doc: task1.id, data: { status: task1.status +1 })) ]
});
```

### Check List
If not relevant to pull request, check off as complete

- [x] All tests passing
- [x] Docs updated with any changes or examples if applicable
- [x] Added tests to ensure new feature(s) work properly

### Relevant Issues
* https://github.com/prescottprue/redux-firestore/issues/323

puppybits commented on src/actions/firestore.js:
```diff
@@ -188,76 +188,87 @@ export function setListener(firebase, dispatch, queryOpts, successCb, errorCb) {
   const meta = getQueryConfig(queryOpts);
 
   // Create listener
-  const unsubscribe = firestoreRef(firebase, meta).onSnapshot(
```
None of the internals change in the onSnapshot. I just move it so uses could optionally add an `includeMetadataChanges` to get listeners to optionally trigger when Firestore writes data to the local IndexDB. This is important when in Li-Fi or offline mode so optimistic data in the cache reducer stays to the minimum.

puppybits commented:
Without the speed & responsiveness boosts from this cache reducer we wouldn't have been able to continue to use RRF. Hopefully it's insightful to others and can help others get better performance out of Firestore.

prescottprue commented:
@puppybits Thanks for the PR! Seems like we should be able to get this in within a new minor since things are still pre v1 - are you aware of any breaking changes? Doesn't seem like there would be any since it is a whole new reducer to handle this case?

For anyone looking for high performance, I have had great results with [reactfire](https://github.com/FirebaseExtended/reactfire) (disclaimer: also a contributor on there) - instead of piping database state into redux, it leave management of that state to the Firebase SDK. Even components which attaching similar queries will only listen on the DB once, [more detail in this medium article](http://bit.ly/3d2XPsS) (also called out in the README of react-redux-firebase)

All of that said, going to look today into what it will take to get this into an upcoming version. Thanks again! 🚀

prescottprue commented on src/utils/mutate.js:
```diff
@@ -0,0 +1,242 @@
+ * @param {string} doc
+ * @returns Boolean
+ */
+const docRef = (firestore, collection, doc) =>
```
Wanted to call out that here we could use a slash path instead of making a util function, this also handles many levels deep:

```js
firestore.doc('collection1/doc1/subcollection/subcollectDoc')
```

puppybits replied:
fixed!

prescottprue commented on src/utils/mutate.js:
```diff
@@ -0,0 +1,242 @@
+
+    operations.writes.forEach((writeFnc) => {
+      const complete = mark('mutate.writeInTransaction:writes');
+      const operation = isFunction(writeFnc) ? writeFnc(reads) : writeFnc;
```
```suggestion
      const operation = typeof writeFnc === 'function' ? writeFnc(reads) : writeFnc;
```

puppybits replied:
fixed!

prescottprue commented on src/utils/mutate.js:
```diff
@@ -0,0 +1,242 @@
+      const coll = firestoreRef(firebase, read);
+      return coll
+        .get()
+        .then((snapshot) => {
```
Since we are already in an async function we should be able to use await instead of mixing promise chaining and async/await

puppybits replied:
Cool. Moved all .then into a async/await.

prescottprue commented on src/utils/mutate.js:
```diff
@@ -0,0 +1,242 @@
+      return coll
+        .get()
+        .then((snapshot) => {
+          if (snapshot.docs.length === 0) return Promise.resolve([]);
```
If this moves up to the async function, then this array could just be returned and it would be the resolve value instead of needing to wrap it (which needs to happen since we are using promise chaining)

puppybits replied:
fixed!

prescottprue commented on src/utils/mutate.js:
```diff
@@ -0,0 +1,242 @@
+
+// ----- FieldValue support -----
+
+const primaryValue = (arr) =>
```
Can these also be named functions to match the other utils? That way the could in theory be used within each other without an issue of ordering one util above/below another

puppybits replied:
The mutate's `atomize` turns JSON data into Firestore writable format. The cacheReducer's `atomize` turns JSON into optimistic update JSON. 

PrimaryValue does need to go first. It decides if it's a custom tuple or not. The other just fall through until one is hit. 

In terms of scope they should only be in the `atomize` function. Assigning the vars outside the function was just to reduce recreating them on every pass. It could also be an inline switch but I thought this was easier to read. 

What do you think is the best place for this code?

prescottprue commented on src/utils/actions.js:
```diff
@@ -32,37 +35,46 @@ export function wrapInDispatch(
     });
+    return Promise.reject(err);
+  };
+
```
If we dynamically pick our promise here, we could just have the success/failure afterwords:

```js
const actionPromise = method === 'mutate' ? mutate(ref, ...args) : ref[method](...args)
return actionPromise
  .then((result) => {
    const successIsObject = isObject(successType);
    // Built action object handling function for custom payload
    const actionObj = {
      type: successIsObject ? successType.type : successType,
      meta,
      payload:
        successIsObject && successType.payload
          ? makePayload(successType, result)
          : { args },
    };
    // Attach preserve to action if it is passed
    if (successIsObject && successType.preserve) {
      actionObj.preserve = successType.preserve;
    }
    // Attach merge to action if it is passed
    if (successIsObject && successType.merge) {
      actionObj.merge = successType.merge;
    }
    dispatch(actionObj);
    return result;
  })
  .catch((err) => {
    dispatch({
      type: errorType,
      meta,
      payload: err,
    });
    return Promise.reject(err);
  })
```

puppybits replied:
Yep that's better. Fixed.

prescottprue commented on src/utils/mutate.js:
```diff
@@ -0,0 +1,242 @@
+import { chunk, cloneDeep, flatten, isFunction, mapValues } from 'lodash';
+import debug from 'debug';
```
The addition of debug should probably happen across more than just this util if we were going to make use of it

Can we pull that out of this change and we can capture it as a separate PR?

puppybits replied:
done.

prescottprue commented on src/reducers/statusReducer.js:
```diff
@@ -71,6 +71,7 @@ export function requestedReducer(state = {}, { type, meta }) {
  * @param {object} action - Object containing the action that was dispatched
  * @param {string} action.type - Type of action that was dispatched
  * @param {string} action.path - Path of action that was dispatched
+ * @param action.meta
```
Can we add the type as well? Seems like it would be object
```suggestion
 * @param {object} action.meta - Metadata for action
```

puppybits replied:
fixed.

prescottprue commented on src/reducers/listenersReducer.js:
```diff
@@ -8,6 +8,8 @@ import { combineReducers } from '../utils/reducers';
  * @param {object} [state={}] - Current listenersById redux state
  * @param {object} action - Object containing the action that was dispatched
  * @param {string} action.type - Type of action that was dispatched
+ * @param action.path
```
```suggestion
 * @param {String} action.path - Path of action
```

puppybits replied:
fixed.

prescottprue commented on src/utils/profiling.js:
```diff
@@ -0,0 +1,38 @@
+import { noop } from 'lodash';
+import debug from 'debug';
```
Great work adding this util 👏

prescottprue commented on src/utils/query.js:
```diff
@@ -166,9 +169,13 @@ export function firestoreRef(firebase, meta) {
+  if (path) ref = ref.collection(path);
   if (collection) ref = ref.collection(collection);
   if (collectionGroup) ref = ref.collectionGroup(collectionGroup);
+  if (id) ref = ref.doc(id);
```
What is the difference between this and `doc`?

puppybits replied:
I really favor homogeneous APIs. Reduce the complexity but reducing the terminology. The mutate is a great example:
```
// single
{collection: 'coll', doc: 'doc', data: {a:1}}
// batch
[{collection: 'coll', doc: 'doc', data: {a:1}}]
// write transaction: 
{write: {collection: 'coll', doc: 'doc', data: {a:1}}}
// read/write transaction:
{read: {myRead: {collection:'coll', doc:'doc'}} 
  write: ({myRead}) => ({collection: 'myCollection', doc: 'myDoc', data: {a:1}})}
```
You notice it's just the same collection/doc/data pattern. Once you see that it's trivial to switch between a type of save.

But when you look at the document references you have a different format. Firestore refs have `id`, and `path` to describe the same thing. So the question is does that extra terminology help communicate a boundary or is it just waste?

Why can't I just take a JSON object that comes out of Firestore and send it back (or a partial of it)?

In the Query Snapshot and DocumentReferenceSnapshots coming out of the listener add `id` and `path` to all data. First it's backward compatible with the data/ordered reducers. Second it's using DocumentReference terms of `id` and `path` to align closely to the official Firestore docs as much as possible. But here's the big question. How do we benefit from query and saving having different terms/API? I don't think we do.

So I support collection/doc as much as possible to make it easy to use the new reducer. The only change devs need to understand is the FieldValue (since I can't reflect then reliably). But by supporting the `id` and `path` for everything it makes the API so much simpler.

```js
const tasks = useSelector(taskSelector);
const {id, path, name} = tasks[0];

// with the current API
dispatch(sendMutate({doc: id, collection: path, data: { name } }));

// with the newer homogeneous API
dispatch(sendMutate({id, path, name}));

// Also mutate function also supports both the current API or the new homogeneous API
// {read: {myRead: {path, id}} 
//  write: ({myRead}) => ({path, id, ...rest})}
```

So in short, `id` and `doc` are interchangeable. `path` and `collection` are mostly interchangeable, but collection supports an object notation where path is the same URL-style. By moving everything to support `id` and `path` the public API is so small, clear and simple.

puppybits commented:
@prescottprue Let me know what you prefer to do with the atomize functions. The rest of the items are addressed. 

In the last week I fixed a few production bugs in our app including adding in per-property reconciliation. This solves the issue when a document changes multiple times before Firestore resolves each change. I'm going to be adding some test to get CC back up.

Also we needed to see data load times to make sure a page load was bundling loads as much as possible and to also see the cost of internal processing for very large data set (like doing 3 populates on 1,500 documents). I added a top-level function to make it easy to pull out stats so we could optimize our code base. 

![image](https://user-images.githubusercontent.com/140163/116733951-a0f26100-a9a1-11eb-92a7-c64214ba219f.png)

puppybits commented:
@prescottprue 

✅ tests pass 
✅ CC pass 
✅ Comments all addressed

Let me know if there's anything else to address.

puppybits commented:
@prescottprue Just pinging you back. Is there anything else you'd like to see or is the PR good to go?

prescottprue commented on README.md:
```diff
@@ -178,36 +189,106 @@ store.firestore.get({ collection: 'cities' }),
+
+For changes that are reflected immeadiately to the user, see the mutate API.
+
+##### set - old api
```
Thanks for updating the README with docs!

I would say that calling this this "old api" leads me to think we should be doing a major version bump here, when I don't really think that is necessary since it isn't breaking this existing function, it just doesn't work with the new stuff.

Since you already put so much work in, I can clean up that wording as part of the release process

prescottprue commented on README.md:
```diff
@@ -216,18 +297,209 @@ store.firestore
+const firestoreDocument = {
+  someString: 'string',
+  someNumber: 1,
+  otherNumber: ['::increment', 5],
```
Is there a reason we can't have this leverage the existing Firestore API through [`firestore.FieldValue.increment`](https://firebase.google.com/docs/reference/js/firebase.firestore.FieldValue#static-increment) instead of a string that becomes a different part of this API? Same with arrayUnion and arrayRemove

If we really can't and the string is needed we should add a clear explanation as to why - the documentation part doesn't have to block the release necessarily, but would actually like to answer that for myself 😆

puppybits replied:
That’s a fair critique. 

FieldValue has an internal property to store values. Right now it’s possible, but ugly, to dig in and get this property. That property is different for each field value. But if they make it private then it would be impossible. If they change the internals of their API it breaks this library. Also passing class instances that are not serializable in a message is frowned upon in Redux’s rule to have action messages serializable. 

I can update the docs to explain fully.

prescottprue commented on docs/rfcs/rfc_firestore_immediate_mode_mutations.md:
```diff
@@ -0,0 +1,316 @@
+Isolating mutations into stages makes the mutations be compositional and serialable.
+Doing so fixed multiple design flaws with redux:
+
+1. Actions finally return to the original deisgn; fully synchronous and serialzable so that they can be replayed if needed.
```
```suggestion
1. Actions finally return to the original design; fully synchronous and serializable so that they can be replayed if needed.
```

prescottprue commented on docs/rfcs/rfc_firestore_immediate_mode_mutations.md:
```diff
@@ -0,0 +1,316 @@
+
+### Process steps
+
+Step 1. Decoding as a global convertion
```
```suggestion
Step 1. Decoding as a global convention
```

prescottprue commented on docs/rfcs/rfc_firestore_immediate_mode_mutations.md:
```diff
@@ -0,0 +1,316 @@
+    writes = operations.writes;
+  }
+
+  // first run all optimistic updates synchronously
```
Wondering if this logic could fit into a redux middleware? I was actually experimenting with [actions which trigger promises in this middleware](https://github.com/prescottprue/redux-firestore/blob/master/src/middleware.js#L83), so it seems like that could be a good place to do an optimistic update with rollback

puppybits replied:
Interesting thought. What are the positive/negatives of using a middleware for this?

In our project we use redux toolkit’s asyncActionCreator in our project. Honestly for us it’s been a bad pattern. Async is the risky code that can be written.

Here’s the reason. Async code, 99% of the time, is asking for data on disk or over the network, right? That means the code will have zero guarantees about what will come back. It’s not idempotent. It’s not encapsulated state. It’s completely outside the coders control. Promises making the result binary of pass/fail makes it more manageable but more branches and more complexity which adds more failure points. 

The better pattern for us has been sync first, async second. Why? Because the coder has full control to guarantee the outcome. You can show the result to the user via a store update then, after all sync operations complete, then kicking off the async, non-deterministic operations. 

The postive of moving it to a middleware that I see could be that it would allow batching multiple actions better. Hopefully that’s a rare case since the mutation API was made to allow the coder to compose multiple operations as one (like a multi-document transaction). 

On the downside it’s mutating a reducer’s store outside of the reducer. Also state is held in memory as a closure (usually through an instance of a promise) making it harder to debug.

prescottprue commented on package.json:
```diff
@@ -24,60 +24,61 @@
+    "sinon": "^10.0.0",
+    "sinon-chai": "^3.6.0",
+    "terser-webpack-plugin": "^5.1.1",
+    "webpack": "^5.33.2",
```
Really nice to see these things updated, but will have to check the difference in webpack 5 output vs 4 to make sure there isn't anything breaking happening

Were any of these dependency updates needed for the functionality you were building or did you just capture it all together? If not, it might be good to get these depedency updates out in a separate PR while we are still discussing the immediate mode API/implementation

prescottprue replied:
Recalling this out here - we should be able to add the cache reducer without these updates right?

prescottprue commented on src/actions/firestore.js:
```diff
@@ -386,6 +402,32 @@ export function runTransaction(firebase, dispatch, transactionPromise) {
+ * @returns {Promise} Resolves with results of update call
+ */
+export function mutate(firebase, dispatch, writes) {
+  return wrapInDispatch(dispatch, {
```
👏

prescottprue commented on src/reducers/errorsReducer.js:
```diff
@@ -35,6 +36,8 @@ function errorsAllIds(state = [], { meta, type }) {
  * @param {object} [state=[]] - Current authError redux state
  * @param {object} action - Object containing the action that was dispatched
  * @param {string} action.type - Type of action that was dispatched
+ * @param action.meta
```
Nit: Can we add types and descriptions to these - both should be object I believe

prescottprue commented on src/utils/actions.js:
```diff
@@ -19,6 +20,8 @@ function makePayload({ payload }, valToPass) {
  * @param {Function} opts.method - Method to call
  * @param {Array} opts.args - Arguments to call method with
  * @param {Array} opts.types - Action types array ([BEFORE, SUCCESS, FAILURE])
+ * @param opts.ref
```
Same here

prescottprue commented on src/utils/mutate.js:
```diff
@@ -0,0 +1,249 @@
+ * @returns Boolean
+ */
+const docRef = (firestore, collection, doc) =>
+  firestore.doc(`${collection}/${doc}`);
```
Is there a reason to make this a separate function if we are just calling the existing method?

puppybits replied:
What happens if/when the firestore API changes? It can be changed just here (because it wraps the library) or change it in multiple places. 

That’s the reason I tried to keep a call to Firestore calls to a minimum. It makes it easier to support when new versions come out. Also a backward compatible of multiple versions is easier because it’s just a single if/else here.

prescottprue commented on src/utils/mutate.js:
```diff
@@ -0,0 +1,249 @@
+
+const arrayUnion = (firebase, key, val) => {
+  if (key !== '::arrayUnion') return null;
+  return firebase.firestore.FieldValue.arrayUnion(val);
```
This may be the answer to my question as to why we need the string keys - I'm wondering though.... could those maybe just be passed along like so:

```js
// changes requested to your data
const firestoreDocument = {
  otherNumber: firebase.firestore.FieldValue.increment(5)
}
```

Not sure it is even necessary, but if they needed to be dynamic to the val, we could maybe use a function like:

```js
// changes requested to your data
const firestoreDocument = {
  otherNumber: (val) => firebase.firestore.FieldValue.increment(val),
}
```

prescottprue replied:
Ahhh.... I think I finally see... Is it because these are going into the store and we need those to remain as plain objects?

puppybits replied:
Right. The value is needed to have the optimistic value in the store. But inspecting the inside of a FieldValue isn’t good. If/when Firestore changes the internals of their code the hacky way to get the value would be different. 

I’ll clean up the docs so this is clear to consumers of the library.

prescottprue commented:
@puppybits Thanks for the ping - looked back and had a few questions. This is looking pretty good to go, just want to make sure I fully understand all the decisions and that the docs are clean

Great work with all of this by the way 🥇

prescottprue commented on README.md:
```diff
@@ -34,689 +34,86 @@ npm install --save react-redux-firebase
 
 [react-redux-firebase](https://github.com/prescottprue/react-redux-firebase) provides [`withFirestore`](http://react-redux-firebase.com/docs/api/withFirestore.html) and [`firestoreConnect`](http://react-redux-firebase.com/docs/api/firestoreConnect.html) higher order components, which handle automatically calling `redux-firestore` internally based on component's lifecycle (i.e. mounting/un-mountin ...
 
-## Use
```
Was about to merge, but was hoping to make this a minor (since there there are already alpha versions of v1 with proposed changes). Because of that we we should probably keep this normal use section for folks which aren't using the cache reducer

prescottprue commented on package.json:
```diff
@@ -24,60 +24,61 @@
-    "immer": "8.0.1",
-    "lodash": "^4.17.20",
+    "debug": "^4.3.1",
+    "immer": "9.0.1",
```
There is a major version change to immer here - was that the intention? To keep this to just a minor version change we will need to understand the impact of this update

prescottprue commented:
Been looking to try to get this merged in, but in order to release it as part of a minor version instead of a major (there is already an alpha and proposed API change for v1) we will need to address some stuff (such as preserving the Use section of the readme, and not updating major versions of libraries unless needed). I will try to get to some of this, but if you are able to make the changes we should be able to get this in - alternatively we can point your changes at the `alpha` branch where v1 alpha currently lives ([mostly a change to the way data is stored in state and how selectors work](https://github.com/prescottprue/redux-firestore/wiki/v1.0.0-Roadmap))

prescottprue commented:
Thanks so much for these amazing changes, the tests, and keep an eye on keeping changes backwards compatible - apologies it took so long to get these in. Cycles should be shorter moving forward

prescottprue commented:
I've noticed a few things which I addressed including eslint warnings/errors.

Working on cleaning up in the [v0.16.0 PR](https://github.com/prescottprue/redux-firestore/pull/366/files) before releasing which includes
* adding back removed reducers - these can be deprecated if we really think they should be removed
* removing the performance util - want to make sure we can have this in without impacting bundle size as much for production build
* removing debug - adding a debug/log util was planned to happen all at once after a bit of research. Debug dependency adds size - we can do this separately (your changes which log whats happening clearly will be revived once this is done) 

One of the new tests is failing though, so I'll need to debug a bit more: https://github.com/prescottprue/redux-firestore/runs/4296991419?check_suite_focus=true#step:6:423
